### ISSUE TITLE
New curl class / httpinterface breaking new logins

### ISSUE THREAD

**Issue by jpmoney5065** [mgp25/Instagram-API#1000]
The new curl class code is giving me `login_required` message when trying to log in new accounts that I don't have cookies for.

- All settings are being set in my mysql table except for the cookies column.
- Previously logged in accounts working fine.

Test code:
```
$i= new \InstagramAPI\Instagram();
$i->setUser('xxxxx','xxxxx');
$i->setProxy('xxx',xx,'xxx','xxx');

try{
    $i->login(true);
}catch(\InstagramAPI\InstagramException $e){
    echo $e->getMessage();
    echo $e->getTraceAsString();
}
```
Stack Trace:
```
#0 mgp25/instagram-php/src/Instagram.php: InstagramAPI\Request->getResponse(Object(InstagramAPI\TimelineFeedResponse))
#1 mgp25/instagram-php/src/Instagram.php: InstagramAPI\Instagram->timelineFeed()
#2 TEST_CODE.php: InstagramAPI\Instagram->login(true)
#3 {main}
```

Http prop of Instagram Object:
```
[http] => InstagramAPI\HttpInterface Object
        (
            [parent:protected] => InstagramAPI\Instagram Object
 *RECURSION*
            [userAgent:protected] => Instagram 10.8.0 Android (18/4.3; 320dpi; 720x1280; Smartfren; AD688G; Smartfren Andromax AD688G; qcom; en_US)
            [verifyPeer:protected] => 1
            [verifyHost:protected] => 2
            [proxy] => Array
                (
                    [host] => xxx.xxxx.xx.xxx
                    [port] => 80
                    [username] => xxxx
                    [password] => xxx
                )

            [outputInterface] => 
            [curl:InstagramAPI\HttpInterface:private] => Curl\Curl Object
                (
                    [curl] => Resource id #54
                    [id] => 58b79124c0a618.83961214
                    [error] => 1
                    [errorCode] => 400
                    [errorMessage] => HTTP/1.1 400 Bad Request
                    [curlError] => 
                    [curlErrorCode] => 0
                    [curlErrorMessage] => 
                    [httpError] => 1
                    [httpStatusCode] => 400
                    [httpErrorMessage] => HTTP/1.1 400 Bad Request
                    [baseUrl] => https://i.instagram.com/api/v1/feed/timeline?rank_token=4668798610_61263d54-ab67-499f-bdc0-0fe350e9ac57&ranked_content=true
                    [url] => https://i.instagram.com/api/v1/feed/timeline?rank_token=4668798610_61263d54-ab67-499f-bdc0-0fe350e9ac57&ranked_content=true
                    [requestHeaders] => Curl\CaseInsensitiveArray Object
                        (
                            [data:Curl\CaseInsensitiveArray:private] => Array
                                (
                                    [request-line] => GET /api/v1/feed/timeline/?rank_token=4668798610_61263d54-ab67-499f-bdc0-0fe350e9ac57&ranked_content=true HTTP/1.1
                                    [user-agent] => Instagram 10.8.0 Android (18/4.3; 320dpi; 720x1280; Smartfren; AD688G; Smartfren Andromax AD688G; qcom; en_US)
                                    [host] => i.instagram.com
                                    [connection] => close
                                    [accept] => */*
                                    [accept-encoding] => gzip, deflate, sdch
                                    [x-ig-capabilities] => 3ToAAA==
                                    [x-ig-connection-type] => WIFI
                                    [x-ig-connection-speed] => 3017kbps
                                    [x-fb-http-engine] => Liger
                                    [content-type] => application/x-www-form-urlencoded; charset=UTF-8
                                    [accept-language] => en-US
                                )

                            [keys:Curl\CaseInsensitiveArray:private] => Array
                                (
                                    [request-line] => Request-Line
                                    [user-agent] => User-Agent
                                    [host] => Host
                                    [connection] => Connection
                                    [accept] => Accept
                                    [accept-encoding] => Accept-Encoding
                                    [x-ig-capabilities] => X-IG-Capabilities
                                    [x-ig-connection-type] => X-IG-Connection-Type
                                    [x-ig-connection-speed] => X-IG-Connection-Speed
                                    [x-fb-http-engine] => X-FB-HTTP-Engine
                                    [content-type] => Content-Type
                                    [accept-language] => Accept-Language
                                )

                        )

                    [responseHeaders] => Curl\CaseInsensitiveArray Object
                        (
                            [data:Curl\CaseInsensitiveArray:private] => Array
                                (
                                    [status-line] => HTTP/1.1 400 Bad Request
                                    [vary] => Cookie, Accept-Language
                                    [content-language] => en
                                    [cache-control] => private, no-cache, no-store, must-revalidate
                                    [expires] => Sat, 01 Jan 2000 00:00:00 GMT
                                    [content-type] => application/json
                                    [date] => Thu, 02 Mar 2017 03:27:33 GMT
                                    [pragma] => no-cache
                                    [set-cookie] => csrftoken=Vn45tpe4ZZ8fegnLzmniOVdYdAtFDfPd; expires=Thu, 01-Mar-2018 03:27:33 GMT; Max-Age=31449600; Path=/; Secure,mid=WLeRJQABAAHpGwe1EAfhVG5Cw_DT; expires=Wed, 25-Feb-2037 03:27:33 GMT; Max-Age=630720000; Path=/
                                    [connection] => close
                                    [content-length] => 47
                                )

                            [keys:Curl\CaseInsensitiveArray:private] => Array
                                (
                                    [status-line] => Status-Line
                                    [vary] => Vary
                                    [content-language] => Content-Language
                                    [cache-control] => Cache-Control
                                    [expires] => Expires
                                    [content-type] => Content-Type
                                    [date] => Date
                                    [pragma] => Pragma
                                    [set-cookie] => Set-Cookie
                                    [connection] => Connection
                                    [content-length] => Content-Length
                                )

                        )

                    [rawResponseHeaders] => HTTP/1.1 200 Connection established

HTTP/1.1 301 Moved Permanently
Date: Thu, 02 Mar 2017 03:27:32 GMT
Content-Language: en
Location: https://i.instagram.com/api/v1/feed/timeline/?rank_token=4668798610_61263d54-ab67-499f-bdc0-0fe350e9ac57&ranked_content=true
Content-Type: text/html; charset=utf-8
Cache-Control: private, no-cache, no-store, must-revalidate
Expires: Sat, 01 Jan 2000 00:00:00 GMT
Vary: Accept-Language
Pragma: no-cache
Connection: close
Content-Length: 0

HTTP/1.1 200 Connection established

HTTP/1.1 400 Bad Request
Vary: Cookie, Accept-Language
Content-Language: en
Cache-Control: private, no-cache, no-store, must-revalidate
Expires: Sat, 01 Jan 2000 00:00:00 GMT
Content-Type: application/json
Date: Thu, 02 Mar 2017 03:27:33 GMT
Pragma: no-cache
Set-Cookie: csrftoken=Vn45tpe4ZZ8fegnLzmniOVdYdAtFDfPd; expires=Thu, 01-Mar-2018 03:27:33 GMT; Max-Age=31449600; Path=/; Secure
Set-Cookie: mid=WLeRJQABAAHpGwe1EAfhVG5Cw_DT; expires=Wed, 25-Feb-2037 03:27:33 GMT; Max-Age=630720000; Path=/
Connection: close
Content-Length: 47


                    [responseCookies] => Array
                        (
                            [csrftoken] => Vn45tpe4ZZ8fegnLzmniOVdYdAtFDfPd
                            [mid] => WLeRJQABAAHpGwe1EAfhVG5Cw_DT
                        )

                    [response] => stdClass Object
                        (
                            [status] => fail
                            [message] => login_required
                        )

                    [rawResponse] => {"status": "fail", "message": "login_required"}
                    [beforeSendFunction] => 
                    [downloadCompleteFunction] => 
                    [successFunction] => 
                    [errorFunction] => 
                    [completeFunction] => 
                    [fileHandle] => 
                    [cookies:Curl\Curl:private] => Array
                        (
                        )

                    [headers:Curl\Curl:private] => Curl\CaseInsensitiveArray Object
                        (
                            [data:Curl\CaseInsensitiveArray:private] => Array
                                (
                                    [connection] => close
                                    [accept] => */*
                                    [accept-encoding] => gzip, deflate, sdch
                                    [x-ig-capabilities] => 3ToAAA==
                                    [x-ig-connection-type] => WIFI
                                    [x-ig-connection-speed] => 3017kbps
                                    [x-fb-http-engine] => Liger
                                    [content-type] => application/x-www-form-urlencoded; charset=UTF-8
                                    [accept-language] => en-US
                                )

                            [keys:Curl\CaseInsensitiveArray:private] => Array
                                (
                                    [connection] => Connection
                                    [accept] => Accept
                                    [accept-encoding] => Accept-Encoding
                                    [x-ig-capabilities] => X-IG-Capabilities
                                    [x-ig-connection-type] => X-IG-Connection-Type
                                    [x-ig-connection-speed] => X-IG-Connection-Speed
                                    [x-fb-http-engine] => X-FB-HTTP-Engine
                                    [content-type] => Content-Type
                                    [accept-language] => Accept-Language
                                )

                        )

                    [options:Curl\Curl:private] => 
                    [jsonDecoder:Curl\Curl:private] => 
                    [jsonPattern:Curl\Curl:private] => /^(?:application|text)\/(?:[a-z]+(?:[\.-][0-9a-z]+){0,}[\+\.]|x-)?json(?:-[a-z]+)?/i
                    [xmlDecoder:Curl\Curl:private] => 
                    [xmlPattern:Curl\Curl:private] => ~^(?:text/|application/(?:atom\+|rss\+)?)xml~i
                    [defaultDecoder:Curl\Curl:private] => 
                )

        )
```

**Comment by SteveJobzniak** [comment]
Hi. HttpInterface is being rewritten to Guzzle instead.

Also be aware that only release-tagged versions are considered stable. You are using the in-development master/head version, so bugs like this are expected. Especially since the HttpInterface rewrite commits all have "(unfinished)" in the title. They are not finished yet. ;-)

For stable tags, go to the "Releases" tab for this project.

**Comment by jpmoney5065** [comment]
@SteveJobzniak I understand, just giving feedback on what was pushed out. 

Happy about guzzle!

**Comment by SteveJobzniak** [comment]
@jpmoney5065 Gotcha. Well it'll all be great soon. ;-) Guzzle will also bring much better performance *and* properly thrown exceptions for all HTTP errors. Which gives the caller unprecedented control of what to do (try{} catch{}) whenever Instagram couldn't be reached! It's going to require some adapting from users but it's important to finally have a modern, correct Exception-based way of handling errors.

I am currently guiding mgp about he rewrite (since I know Guzzle) and it's going very well.

**Comment by jpmoney5065** [comment]
Been trying to learn guzzle myself the last week or so, so happy it will be included. I look forward to pooling requests!!

**Comment by SteveJobzniak** [comment]
@jpmoney5065 Hehe, pooling in our case will "just" mean having a pool of reusable cURL instances instead of constantly creating/destroying new instances. Guzzle ensures that it reuses cURL instances to reduce overhead, and it also keeps HTTPS connections alive to instagram so that each subsequent API request can instantly be performed.

If you're thinking about the _other_ kind of pooling, a Guzzle Request Pool(), that won't be used in this library. Such pools are mainly used for concurrent file downloads. They make no sense for sequential API requests. ;-)

**Comment by jpmoney5065** [comment]
@SteveJobzniak  ahh ok.. I am currently using Guzzle Request Pool in conjunction with this API though. 

I do requests from many accounts at once and need to do them at the same time.

**Comment by SteveJobzniak** [comment]
@jpmoney5065 Such specialized needs definitely exist but they aren't normal. We would have to rewrite EVERYTHING in this livrary to support being logged in to multiple accounts at the same time in a single Instagram-API class instance, and we would have to rewrite every API function to support pooling of requests and we would have to rewrite every API function to use deferred results via callbacks (because we could no longer instantly return an answer from any API functions). And then we'd have to add a function to run the pool, so that all requests are performed and trigger their callbacks. In other words forget it. It would destroy the library and make it a mess to use. ;-)

**Comment by SteveJobzniak** [comment]
You could consider running multiple copies of your script in parallel using separate PHP instances to process multiple accounts simultaneously, one account per PHP instance

**Comment by jpmoney5065** [comment]
Yeah I've been using Guzzle request pools to make a bunch of concurrent requests to my script. Then the script will cycle through users using one instance of the API,
